shell/browser/api/electron_api_web_contents.cc

```diff
@@ -2492,6 +2492,58 @@ content::NavigationEntry* WebContents::GetNavigationEntryAtIndex(
   return web_contents()->GetController().GetEntryAtIndex(index);
 }
 
+bool WebContents::DeleteNavigationEntryAtIndex(int index) {
+  if (index < 0 || index >= GetHistoryLength())
+    return false;
+
+  return web_contents()->GetController().RemoveEntryAtIndex(index);
+}
+
+std::vector<content::NavigationEntry*> WebContents::GetHistory() const {
+  const int history_length = GetHistoryLength();
+  auto& controller = web_contents()->GetController();
+
+  // If the history is empty, it contains only one entry and that is
+  // "InitialEntry"
+  if (history_length == 1 && controller.GetEntryAtIndex(0)->IsInitialEntry())
+    return std::vector<content::NavigationEntry*>();
+
+  std::vector<content::NavigationEntry*> history;
+  history.reserve(history_length);
+
+  for (int i = 0; i < history_length; i++)
+    history.push_back(controller.GetEntryAtIndex(i));
+
+  return history;
+}
+
+bool WebContents::ReplaceHistory(
+    const std::vector<gin_helper::Dictionary>& new_history,
+    int index) {
+  std::vector<std::unique_ptr<content::NavigationEntry>> history;
+  history.reserve(new_history.size());
+
+  for (const gin_helper::Dictionary& history_item : new_history) {
+    auto entry = content::NavigationEntry::Create();
+    entry->SetIsOverridingUserAgent(true);
+    std::u16string url, title;
+    if (history_item.Get("url", &url))
+      entry->SetURL(GURL(url));
+    if (history_item.Get("title", &title))
+      entry->SetTitle(title);
+
+    history.push_back(std::move(entry));
+  }
+
+  auto& controller = web_contents()->GetController();
+  if (!controller.CanPruneAllButLastCommitted())
+    return false;
+
+  controller.PruneAllButLastCommitted();
+  controller.Restore(index, content::RestoreType::kRestored, &history);
```

Comment:
My gut says this API is a security concern in general and correct me if you disagree. We are allowing the override of the navigation history without any input validation. Anyone could potentially inject malicious scripts the the url or page titles. We are also circumventing user's privacy/ tracking mechanisms if they didn't explicitly intend to navigate to a particular entry.

Reply:
Given this is an API provided in the browser process, I'd place security concerns at the same level as `WebContents.loadURL`. Modifying the URL or history of a WebContents should be a security concern of the application author which we'll only be able to advise on from a framework author's perspective.

Our security recommendations are to [limit navigation](https://www.electronjs.org/docs/latest/tutorial/security#13-disable-or-limit-navigation) based on the `will-navigate` event. Tests should be added to verify that this is still possible when history is replaced.